File: packages/py-moose-lib/moose_lib/dmv2/ingest_pipeline.py
"""
Ingestion Pipeline definitions for Moose Data Model v2 (dmv2).

This module provides classes for defining and configuring complete ingestion pipelines,
which combine tables, streams, and ingestion APIs into a single cohesive unit.
"""
from typing import Any, Optional, Generic, TypeVar
from pydantic import BaseModel

from .types import TypedMooseResource, T
from .olap_table import OlapTable, OlapConfig
from .stream import Stream, StreamConfig, DeadLetterQueue
from .ingest_api import IngestApi, IngestConfig, IngestConfigWithDestination

class IngestPipelineConfig(BaseModel):
    """Configuration for creating a complete ingestion pipeline.

    Defines which components (table, stream, ingest API) should be created.
    Set a component to `True` for default settings, `False` to disable, or provide
    a specific config object (`OlapConfig`, `StreamConfig`, `IngestConfig`).

    Attributes:
        table: Configuration for the OLAP table component.
        stream: Configuration for the stream component.
        ingest: Configuration for the ingest API component.
        dead_letter_queue: Configuration for the dead letter queue component.
        version: Optional version string applied to all created components.
        metadata: Optional metadata for the ingestion pipeline.
    """
    table: bool | OlapConfig = True
    stream: bool | StreamConfig = True
    ingest: bool | IngestConfig = True
    dead_letter_queue: bool | StreamConfig = True
    version: Optional[str] = None
    metadata: Optional[dict] = None

class IngestPipeline(TypedMooseResource, Generic[T]):
    """Creates and configures a linked Table, Stream, and Ingest API pipeline.

    Simplifies the common pattern of ingesting data through an API, processing it
    in a stream, and storing it in a table.

    Args:
        name: The base name used for all created components (table, stream, API).
        config: Specifies which components to create and their configurations.
        t: The Pydantic model defining the data schema for all components
           (passed via `IngestPipeline[MyModel](...)`).

    Attributes:
        table: The created `OlapTable` instance, if configured.
        stream: The created `Stream` instance, if configured.
        ingest_api: The created `IngestApi` instance, if configured.
        dead_letter_queue: The created `DeadLetterQueue` instance, if configured.
        columns (Columns[T]): Helper for accessing data field names safely.
        name (str): The base name of the pipeline.
        model_type (type[T]): The Pydantic model associated with this pipeline.
    """
    table: Optional[OlapTable[T]] = None
    stream: Optional[Stream[T]] = None
    ingest_api: Optional[IngestApi[T]] = None
    dead_letter_queue: Optional[DeadLetterQueue[T]] = None
    metadata: Optional[dict] = None

    def get_table(self) -> OlapTable[T]:
        """Retrieves the pipeline's OLAP table component.

        Raises:
            ValueError: If the table was not configured for this pipeline.

        Returns:
            The `OlapTable` instance.
        """
        if self.table is None:
            raise ValueError("Table was not configured for this pipeline")
        return self.table

    def get_stream(self) -> Stream[T]:
        """Retrieves the pipeline's stream component.

        Raises:
            ValueError: If the stream was not configured for this pipeline.

        Returns:
            The `Stream` instance.
        """
        if self.stream is None:
            raise ValueError("Stream was not configured for this pipeline")
        return self.stream

    def get_dead_letter_queue(self) -> Stream[T]:
        """Retrieves the pipeline's dead letter queue.

        Raises:
            ValueError: If the dead letter queue was not configured for this pipeline.

        Returns:
            The `Stream` instance.
        """
        if self.dead_letter_queue is None:
            raise ValueError("DLQ was not configured for this pipeline")
        return self.dead_letter_queue

    def get_ingest_api(self) -> IngestApi[T]:
        """Retrieves the pipeline's Ingestion API component.

        Raises:
            ValueError: If the Ingest API was not configured for this pipeline.

        Returns:
            The `IngestApi` instance.
        """
        if self.ingest_api is None:
            raise ValueError("Ingest API was not configured for this pipeline")
        return self.ingest_api

    def __init__(self, name: str, config: IngestPipelineConfig, **kwargs):
        super().__init__()
        self._set_type(name, self._get_type(kwargs))
        
        # Initialize metadata
        self.metadata = config.metadata or {}
        
        # Apply version to all components if provided
        version = config.version
        
        # Create table if configured
        if config.table:
            table_config = OlapConfig() if config.table is True else config.table
            if version:
                table_config.version = version
            table_config.metadata = self.metadata.copy()
            self.table = OlapTable(name, table_config, t=self._t)
        
        # Create stream if configured
        if config.stream:
            stream_config = StreamConfig() if config.stream is True else config.stream
            if config.table and stream_config.destination is not None:
                raise ValueError("The destination of the stream should be the table created in the IngestPipeline")
            stream_config.destination = self.table
            if version:
                stream_config.version = version
            stream_config.metadata = self.metadata.copy()
            self.stream = Stream(name, stream_config, t=self._t)
        
        # Create dead letter queue if configured
        if config.dead_letter_queue:
            dlq_config = StreamConfig() if config.dead_letter_queue is True else config.dead_letter_queue
            if version:
                dlq_config.version = version
            dlq_config.metadata = self.metadata.copy()
            self.dead_letter_queue = DeadLetterQueue(f"{name}DeadLetterQueue", dlq_config, t=self._t)
        
        # Create ingest API if configured
        if config.ingest:
            if self.stream is None:
                raise ValueError("Ingest API needs a stream to write to.")
            
            # Create ingest config with version
            ingest_config = {}
            if isinstance(config.ingest, IngestConfig):
                ingest_config = config.ingest.model_dump()
            
            # Set destination, version, and DLQ
            ingest_config["destination"] = self.stream
            if version:
                ingest_config["version"] = version
            if self.dead_letter_queue:
                ingest_config["dead_letter_queue"] = self.dead_letter_queue
            
            # Add metadata
            ingest_config["metadata"] = self.metadata.copy()
            
            # Create the ingest API
            ingest_config_obj = IngestConfigWithDestination(**ingest_config)
            self.ingest_api = IngestApi(name, ingest_config_obj, t=self._t)

File: packages/ts-moose-lib/src/dmv2/sdk/ingestPipeline.ts
import { IJsonSchemaCollection } from "typia";
import { TypedBase, TypiaValidators } from "../typedBase";
import { Column } from "../../dataModels/dataModelTypes";
import {
  DeadLetterModel,
  DeadLetterQueue,
  Stream,
  StreamConfig,
} from "./stream";
import { OlapConfig, OlapTable } from "./olapTable";
import { IngestApi, IngestConfig } from "./ingestApi";
import { LifeCycle } from "./lifeCycle";
import { ClickHouseEngines } from "../../blocks/helpers";

/**
 * Configuration options for a complete ingestion pipeline, potentially including an Ingest API, a Stream, and an OLAP Table.
 *
 * @template T The data type of the records being ingested.
 *
 * @example
 * ```typescript
 * // Simple pipeline with all components enabled
 * const pipelineConfig: IngestPipelineConfig<UserData> = {
 *   table: true,
 *   stream: true,
 *   ingest: true
 * };
 *
 * // Advanced pipeline with custom configurations
 * const advancedConfig: IngestPipelineConfig<UserData> = {
 *   table: { orderByFields: ['timestamp', 'userId'], engine: ClickHouseEngines.ReplacingMergeTree },
 *   stream: { parallelism: 4, retentionPeriod: 86400 },
 *   ingest: true,
 *   version: '1.2.0',
 *   metadata: { description: 'User data ingestion pipeline' }
 * };
 * ```
 */
export type IngestPipelineConfig<T> = {
  /**
   * Configuration for the OLAP table component of the pipeline.
   *
   * - If `true`, a table with default settings is created.
   * - If an `OlapConfig` object is provided, it specifies the table's configuration.
   * - If `false`, no OLAP table is created.
   *
   * @default false
   */
  table: boolean | OlapConfig<T>;

  /**
   * Configuration for the stream component of the pipeline.
   *
   * - If `true`, a stream with default settings is created.
   * - If a partial `StreamConfig` object (excluding `destination`) is provided, it specifies the stream's configuration.
   * - The stream's destination will automatically be set to the pipeline's table if one exists.
   * - If `false`, no stream is created.
   *
   * @default false
   */
  stream: boolean | Omit<StreamConfig<T>, "destination">;

  /**
   * Configuration for the ingest API component of the pipeline.
   *
   * - If `true`, an ingest API with default settings is created.
   * - If a partial `IngestConfig` object (excluding `destination`) is provided, it specifies the API's configuration.
   * - The API's destination will automatically be set to the pipeline's stream if one exists.
   * - If `false`, no ingest API is created.
   *
   * **Note:** Requires a stream to be configured when enabled.
   *
   * @default false
   */
  ingest: boolean | Omit<IngestConfig<T>, "destination">;

  /**
   * Configuration for the dead letter queue of the pipeline.
   * If `true`, a dead letter queue with default settings is created.
   * If a partial `StreamConfig` object (excluding `destination`) is provided, it specifies the dead letter queue's configuration.
   * The API's destination will automatically be set to the pipeline's stream if one exists.
   * If `false` or `undefined`, no dead letter queue is created.
   */
  deadLetterQueue?: boolean | StreamConfig<DeadLetterModel>;

  /**
   * An optional version string applying to all components (table, stream, ingest) created by this pipeline configuration.
   * This version will be used for schema versioning and component identification.
   *
   * @example "v1.0.0", "2023-12", "prod"
   */
  version?: string;

  /**
   * Optional metadata for the pipeline.
   */
  metadata?: {
    /** Human-readable description of the pipeline's purpose */
    description?: string;
  };

  /** Determines how changes in code will propagate to the resources. */
  lifeCycle?: LifeCycle;
};

/**
 * Represents a complete ingestion pipeline, potentially combining an Ingest API, a Stream, and an OLAP Table
 * under a single name and configuration. Simplifies the setup of common ingestion patterns.
 *
 * This class provides a high-level abstraction for creating data ingestion workflows that can include:
 * - An HTTP API endpoint for receiving data
 * - A streaming component for real-time data processing
 * - An OLAP table for analytical queries
 *
 * @template T The data type of the records flowing through the pipeline. This type defines the schema for the
 *             Ingest API input, the Stream messages, and the OLAP Table rows.
 *
 * @example
 * ```typescript
 * // Create a complete pipeline with all components
 * const userDataPipeline = new IngestPipeline('userData', {
 *   table: true,
 *   stream: true,
 *   ingest: true,
 *   version: '1.0.0',
 *   metadata: { description: 'Pipeline for user registration data' }
 * });
 *
 * // Create a pipeline with only table and stream
 * const analyticsStream = new IngestPipeline('analytics', {
 *   table: { orderByFields: ['timestamp'], engine: ClickHouseEngines.ReplacingMergeTree },
 *   stream: { parallelism: 8, retentionPeriod: 604800 },
 *   ingest: false
 * });
 * ```
 */
export class IngestPipeline<T> extends TypedBase<T, IngestPipelineConfig<T>> {
  /**
   * The OLAP table component of the pipeline, if configured.
   * Provides analytical query capabilities for the ingested data.
   * Only present when `config.table` is not `false`.
   */
  table?: OlapTable<T>;

  /**
   * The stream component of the pipeline, if configured.
   * Handles real-time data flow and processing between components.
   * Only present when `config.stream` is not `false`.
   */
  stream?: Stream<T>;

  /**
   * The ingest API component of the pipeline, if configured.
   * Provides HTTP endpoints for data ingestion.
   * Only present when `config.ingest` is not `false`.
   */
  ingestApi?: IngestApi<T>;

  /** The dead letter queue of the pipeline, if configured. */
  deadLetterQueue?: DeadLetterQueue<T>;

  /**
   * Creates a new IngestPipeline instance.
   * Based on the configuration, it automatically creates and links the IngestApi, Stream, and OlapTable components.
   *
   * @param name The base name for the pipeline components (e.g., "userData" could create "userData" table, "userData" stream, "userData" ingest API).
   * @param config Optional configuration for the ingestion pipeline.
   *
   * @throws {Error} When ingest API is enabled but no stream is configured, since the API requires a stream destination.
   *
   * @example
   * ```typescript
   * const pipeline = new IngestPipeline('events', {
   *   table: { orderByFields: ['timestamp'], engine: ClickHouseEngines.ReplacingMergeTree },
   *   stream: { parallelism: 2 },
   *   ingest: true
   * });
   * ```
   */
  constructor(name: string, config: IngestPipelineConfig<T>);

  /**
   * Internal constructor used by the framework for advanced initialization.
   *
   * @internal
   * @param name The base name for the pipeline components.
   * @param config Configuration specifying which components to create and their settings.
   * @param schema JSON schema collection for type validation.
   * @param columns Column definitions for the data model.
   */
  constructor(
    name: string,
    config: IngestPipelineConfig<T>,
    schema: IJsonSchemaCollection.IV3_1,
    columns: Column[],
    validators: TypiaValidators<T>,
  );

  constructor(
    name: string,
    config: IngestPipelineConfig<T>,
    schema?: IJsonSchemaCollection.IV3_1,
    columns?: Column[],
    validators?: TypiaValidators<T>,
  ) {
    super(name, config, schema, columns, validators);

    // Create OLAP table if configured
    if (config.table) {
      const tableConfig: OlapConfig<T> =
        typeof config.table === "object" ?
          {
            ...config.table,
            ...(config.version && { version: config.version }),
          }
        : {
            lifeCycle: config.lifeCycle,
            engine: ClickHouseEngines.MergeTree,
            ...(config.version && { version: config.version }),
          };
      this.table = new OlapTable(
        name,
        tableConfig,
        this.schema,
        this.columnArray,
        this.validators,
      );
    }

    // Create stream if configured, linking it to the table as destination
    if (config.stream) {
      const streamConfig = {
        destination: this.table,
        ...(typeof config.stream === "object" ?
          config.stream
        : { lifeCycle: config.lifeCycle }),
        ...(config.version && { version: config.version }),
      };
      this.stream = new Stream(
        name,
        streamConfig,
        this.schema,
        this.columnArray,
      );
      // Set pipeline parent reference for internal framework use
      (this.stream as any).pipelineParent = this;
    }

    if (config.deadLetterQueue) {
      const streamConfig = {
        destination: undefined,
        ...(typeof config.deadLetterQueue === "object" ?
          config.deadLetterQueue
        : {}),
        ...(config.version && { version: config.version }),
      };
      this.deadLetterQueue = new DeadLetterQueue<T>(
        `${name}DeadLetterQueue`,
        streamConfig,
        validators!.assert!,
      );
    }

    // Create ingest API if configured, requiring a stream as destination
    if (config.ingest) {
      if (!this.stream) {
        throw new Error("Ingest API needs a stream to write to.");
      }

      const ingestConfig = {
        destination: this.stream,
        deadLetterQueue: this.deadLetterQueue,
        ...(typeof config.ingest === "object" ? config.ingest : {}),
        ...(config.version && { version: config.version }),
      };
      this.ingestApi = new IngestApi(
        name,
        ingestConfig,
        this.schema,
        this.columnArray,
      );
      // Set pipeline parent reference for internal framework use
      (this.ingestApi as any).pipelineParent = this;
    }
  }
}
